File: android/app/src/main/java/com/laneshadow/ui/components/molecules/CardSkeleton.kt
package com.laneshadow.ui.components.molecules

import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * CardSkeleton molecule component
 *
 * Loading placeholder with pulse animation for card content.
 * Mimics the layout of RouteAttachmentCard: badge row, title, description, stats.
 *
 * @param compact Show compact variant - reduces padding and gaps (default: false)
 * @param showBestBadge Show best badge placeholder (default: true)
 * @param showWeatherBadge Show weather badge placeholder (default: true)
 * @param modifier Modifier for the component
 * @param testId Test ID for UI testing
 */
@Composable
fun CardSkeleton(
    compact: Boolean = false,
    showBestBadge: Boolean = true,
    showWeatherBadge: Boolean = true,
    modifier: Modifier = Modifier,
    testId: String? = null,
) {
    val theme = LocalLaneShadowTheme.current

    // Shimmer pulse animation (same pattern as Skeleton atom)
    val infiniteTransition = rememberInfiniteTransition(label = "card_skeleton_pulse")
    val alpha by infiniteTransition.animateFloat(
        initialValue = 0.4f,
        targetValue = 1f,
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = 1500, easing = FastOutSlowInEasing),
            repeatMode = RepeatMode.Reverse,
        ),
        label = "card_skeleton_alpha",
    )

    // Card styling
    val cardPadding = if (compact) theme.space.md else theme.space.lg
    val gap = if (compact) 6.dp else 10.dp

    Surface(
        modifier = modifier
            .testTag(testId ?: "card-skeleton")
            .semantics {
                contentDescription = "Loading"
            }
            .alpha(alpha),
        shape = RoundedCornerShape(theme.radius.lg),
        color = theme.colors.surface.default,
        border = BorderStroke(1.dp, theme.colors.border.default),
    ) {
        Column(
            modifier = Modifier
                .padding(cardPadding)
                .fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(gap),
        ) {
            // Badge row
            if (showBestBadge || showWeatherBadge) {
                BadgeRow(
                    showBestBadge = showBestBadge,
                    showWeatherBadge = showWeatherBadge,
                    theme = theme,
                )
            }

            // Title placeholder
            SkeletonBar(
                width = 0.7f,
                height = if (compact) 14.dp else 16.dp,
                theme = theme,
                testTag = testId?.let { "$it-title" } ?: "card-skeleton-title",
            )

            // Description placeholders (only in non-compact)
            if (!compact) {
                Column(
                    verticalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    SkeletonBar(
                        width = 1.0f,
                        height = 14.dp,
                        theme = theme,
                        testTag = testId?.let { "$it-description-1" } ?: "card-skeleton-description-1",
                    )
                    SkeletonBar(
                        width = 0.6f,
                        height = 14.dp,
                        theme = theme,
                        testTag = testId?.let { "$it-description-2" } ?: "card-skeleton-description-2",
                    )
                }
            }

            // Stats placeholder
            SkeletonBar(
                width = 0.85f,
                height = if (compact) 11.dp else 13.dp,
                theme = theme,
                modifier = Modifier.padding(top = 2.dp),
                testTag = testId?.let { "$it-stats" } ?: "card-skeleton-stats",
            )
        }
    }
}

/**
 * Badge row skeleton component
 *
 * @param showBestBadge Show best badge placeholder
 * @param showWeatherBadge Show weather badge placeholder
 * @param theme LaneShadow theme values
 */
@Composable
private fun BadgeRow(
    showBestBadge: Boolean,
    showWeatherBadge: Boolean,
    theme: com.laneshadow.theme.LaneShadowThemeValues,
) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        if (showBestBadge) {
            Box(
                modifier = Modifier
                    .background(
                        color = theme.colors.muted.default,
                        shape = RoundedCornerShape(theme.radius.md),
                    )
                    .padding(horizontal = 10.dp, vertical = 4.dp),
            ) {
                SkeletonBar(
                    width = 40.dp,
                    height = 12.dp,
                    theme = theme,
                    testTag = "card-skeleton-best-badge",
                )
            }
        }

        if (showWeatherBadge) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(4.dp),
                modifier = Modifier
                    .background(
                        color = theme.colors.muted.default,
                        shape = RoundedCornerShape(theme.radius.md),
                    )
                    .padding(horizontal = 10.dp, vertical = 4.dp),
            ) {
                // Weather icon circle
                Box(
                    modifier = Modifier
                        .size(14.dp)
                        .background(
                            color = theme.colors.surfaceVariant.default,
                            shape = RoundedCornerShape(theme.radius.sm),
                        )
                        .testTag("card-skeleton-weather-icon"),
                )

                // Weather text bar
                SkeletonBar(
                    width = 40.dp,
                    height = 12.dp,
                    theme = theme,
                    testTag = "card-skeleton-weather-text",
                )
            }
        }
    }
}

/**
 * Skeleton bar component
 *
 * @param width Bar width as fraction of parent or fixed Dp
 * @param height Bar height
 * @param theme LaneShadow theme values
 * @param modifier Modifier for the bar
 * @param testTag Test ID for UI testing
 */
@Composable
private fun SkeletonBar(
    width: Float,
    height: Dp,
    theme: com.laneshadow.theme.LaneShadowThemeValues,
    modifier: Modifier = Modifier,
    testTag: String? = null,
) {
    Box(
        modifier = modifier
            .fillMaxWidth(width)
            .height(height)
            .background(
                color = theme.colors.muted.default,
                shape = RoundedCornerShape(theme.radius.sm),
            )
            .testTag(testTag ?: "skeleton-bar"),
    )
}

/**
 * Skeleton bar component with fixed width
 *
 * @param width Bar width in Dp
 * @param height Bar height
 * @param theme LaneShadow theme values
 * @param modifier Modifier for the bar
 * @param testTag Test ID for UI testing
 */
@Composable
private fun SkeletonBar(
    width: Dp,
    height: Dp,
    theme: com.laneshadow.theme.LaneShadowThemeValues,
    modifier: Modifier = Modifier,
    testTag: String? = null,
) {
    Box(
        modifier = modifier
            .width(width)
            .height(height)
            .background(
                color = theme.colors.muted.default,
                shape = RoundedCornerShape(theme.radius.sm),
            )
            .testTag(testTag ?: "skeleton-bar"),
    )
}

File: ios/LaneShadow/Views/Molecules/CardSkeleton.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Card Skeleton Component

/**
 * Card skeleton loading placeholder molecule
 *
 * Shimmer effect for card content placeholders.
 * Mimics the layout of RouteAttachmentCard: badge row, title, description, stats.
 * Uses pulse opacity animation (1.5s cycle) with easeInOut.
 *
 * ## Design Tokens Used
 * - Colors:
 *   - `theme.colors.surface.default` (card background)
 *   - `theme.colors.muted.default` (skeleton fill)
 *   - `theme.colors.border.default` (border)
 * - Layout:
 *   - Card corner radius: 16pt
 *   - Padding: 16pt (default), 12pt (compact)
 *   - Gap between elements: 10pt (default), 6pt (compact)
 * - Skeleton elements:
 *   - Corner radius: 8pt
 *   - Badge width: 60pt, height: 20pt
 *   - Title height: 16pt (default), 14pt (compact)
 *   - Description height: 14pt
 *   - Stats height: 13pt (default), 11pt (compact)
 *
 * ## Parameters
 * - compact: Show compact variant with reduced padding and gaps
 * - showBestBadge: Show best badge placeholder
 * - showWeatherBadge: Show weather badge placeholder
 *
 * ## Accessibility
 * - Screen reader announces "Loading" via accessibilityLabel
 */
public struct LSCardSkeleton: View {
    @Environment(\.theme) private var theme
    @State private var isPulsing = false

    private let compact: Bool
    private let showBestBadge: Bool
    private let showWeatherBadge: Bool

    public init(
        compact: Bool = false,
        showBestBadge: Bool = true,
        showWeatherBadge: Bool = true
    ) {
        self.compact = compact
        self.showBestBadge = showBestBadge
        self.showWeatherBadge = showWeatherBadge
    }

    // MARK: - Body

    public var body: some View {
        VStack(alignment: .leading, spacing: compact ? 6 : 10) {
            // Badge row
            if showBestBadge || showWeatherBadge {
                HStack(spacing: 8) {
                    if showBestBadge {
                        bestBadgeSkeleton
                    }

                    if showWeatherBadge {
                        weatherBadgeSkeleton
                    }

                    Spacer()
                    statBar(width: 40, height: compact ? 11 : 13)
                }
            }

            // Title placeholder
            skeletonBar(width: 0.6, height: compact ? 14 : 16)

            // Description placeholders (only in non-compact)
            if !compact {
                VStack(alignment: .leading, spacing: 6) {
                    skeletonBar(width: 0.9, height: 14)
                    skeletonBar(width: 0.6, height: 14)
                }
            }

            // Stats placeholder
            statBar(width: 0.85, height: compact ? 11 : 13)
        }
        .padding(compact ? 12 : 16)
        .background(theme.colors.surface.default)
        .clipShape(RoundedRectangle(cornerRadius: 16, style: .continuous))
        .overlay {
            RoundedRectangle(cornerRadius: 16, style: .continuous)
                .stroke(theme.colors.border.default, lineWidth: 1)
        }
        .opacity(isPulsing ? 0.4 : 1.0)
        .animation(.easeInOut(duration: 1.5).repeatForever(autoreverses: true), value: isPulsing)
        .onAppear {
            isPulsing = true
        }
        .accessibilityLabel("Loading")
        .accessibilityElement(children: .ignore)
    }

    // MARK: - Private Views

    private var bestBadgeSkeleton: some View {
        HStack(spacing: 4) {
            skeletonBar(width: 40, height: 12)
        }
        .padding(.horizontal, 10)
        .padding(.vertical, 4)
        .background(
            RoundedRectangle(cornerRadius: theme.radius.lg, style: .continuous)
                .fill(theme.colors.muted.default)
        )
    }

    private var weatherBadgeSkeleton: some View {
        HStack(spacing: 4) {
            // Weather icon circle
            Circle()
                .fill(theme.colors.surfaceVariant.pressed)
                .frame(width: 14, height: 14)

            // Weather text bar
            skeletonBar(width: 40, height: 12)
        }
        .padding(.horizontal, 10)
        .padding(.vertical, 4)
        .background(
            RoundedRectangle(cornerRadius: theme.radius.lg, style: .continuous)
                .fill(theme.colors.muted.default)
        )
    }

    private func skeletonBar(width: CGFloat, height: CGFloat) -> some View {
        RoundedRectangle(cornerRadius: 8, style: .continuous)
            .fill(theme.colors.muted.default)
            .frame(width: width, height: height)
    }

    private func statBar(width: CGFloat, height: CGFloat) -> some View {
        skeletonBar(width: width, height: height)
    }
}

// MARK: - Width Extension for Fractional Widths

private extension CGFloat {
    /// Converts a fractional value (0.0-1.0) to actual width based on a base width
    static func fractional(_ fraction: CGFloat, base: CGFloat = 300) -> CGFloat {
        base * fraction
    }
}

// MARK: - Preview

#Preview("CardSkeleton - Default") {
    VStack(spacing: 16) {
        LSCardSkeleton()
            .frame(width: 320)

        LSCardSkeleton(compact: true)
            .frame(width: 320)

        LSCardSkeleton(showBestBadge: false, showWeatherBadge: false)
            .frame(width: 320)
    }
    .padding()
    .frame(maxWidth: .infinity, maxHeight: .infinity)
    .background(Color.gray.opacity(0.1))
    .laneShadowTheme()
}

#Preview("CardSkeleton - Compact") {
    VStack(spacing: 16) {
        LSCardSkeleton(compact: true)
            .frame(width: 320)

        LSCardSkeleton(compact: true, showBestBadge: false)
            .frame(width: 320)
    }
    .padding()
    .frame(maxWidth: .infinity, maxHeight: .infinity)
    .background(Color.gray.opacity(0.1))
    .laneShadowTheme()
}

#Preview("CardSkeleton - No Badges") {
    LSCardSkeleton(showBestBadge: false, showWeatherBadge: false)
        .frame(width: 320)
        .padding()
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Color.gray.opacity(0.1))
        .laneShadowTheme()
}
